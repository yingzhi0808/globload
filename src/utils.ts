import path from "node:path";

export function normalizePath(p: string): string {
  return path.posix.normalize(p).replace(/\\/g, "/");
}

export function createURL(input: string, base?: string): URL | null {
  if (compareNodeVersions(process.version, "v22.1") >= 0)
    return URL.parse(input, base);

  return new URL(input, base);
}

export function compareNodeVersions(v1: string, v2: string): number {
  const parseVersion = (vStr: string) => {
    const normalized = vStr.startsWith("v") ? vStr.substring(1) : vStr;
    const parts = normalized.split(".").map(Number);
    return {
      major: parts[0] || 0,
      minor: parts[1] || 0,
      patch: parts[2] || 0,
    };
  };

  const ver1 = parseVersion(v1);
  const ver2 = parseVersion(v2);

  if (ver1.major !== ver2.major) {
    return ver1.major > ver2.major ? 1 : -1;
  }
  if (ver1.minor !== ver2.minor) {
    return ver1.minor > ver2.minor ? 1 : -1;
  }
  if (ver1.patch !== ver2.patch) {
    return ver1.patch > ver2.patch ? 1 : -1;
  }
  return 0;
}
