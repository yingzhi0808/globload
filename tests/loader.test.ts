import child_process from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { promisify } from "node:util";
import { afterAll, describe, expect, it } from "vitest";
import { compareNodeVersions } from "../src/utils.ts";

const supportImportAttributes =
	compareNodeVersions(process.version, "v18.20.0") >= 0;

const exec = promisify(child_process.exec);
const registerUrl = pathToFileURL(
	path.resolve(process.cwd(), "dist/register.js"),
).href;

describe("Glob Loader", () => {
	afterAll(async () => {
		await fs.writeFile(mainPath, "");
	});

	let mainPath = path.resolve(process.cwd(), "tests/fixtures/main.js");

	it("should static import lazily with '?glob'", async () => {
		const fileContent = `
			import modules from "./basic/*.js?glob";
			console.log(modules);
			Object.values(modules).forEach(async (module) => {
				console.log(await module());
			});
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should static import eagerly with '?glob&eager'", async () => {
		const fileContent = `
			import modules from "./basic/*.js?glob&eager";
			console.log(modules);
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should dynamic import() lazily with '?glob'", async () => {
		const fileContent = `
			import("./basic/*.js?glob").then((modules) => {
				console.log(modules.default);
				Object.values(modules.default).forEach(async (service) => {
					console.log(await service());
				});
			});	
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should dynamic import() eagerly with '?glob&eager'", async () => {
		const fileContent = `
			import("./basic/*.js?glob&eager").then((modules) => {
				console.log(modules.default);
			});	
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should eagerly load nested (**) modules with '?glob&eager'", async () => {
		const fileContent = `
			import modules from "./nested/**/*.js?glob&eager";
			console.log(modules);
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should eagerly load empty directory as empty object with '?glob&eager'", async () => {
		const fileContent = `
			import modules from "./empty/*.js?glob&eager";
			console.log(modules);
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it("should eagerly load non-existent path as empty object with '?glob&eager'", async () => {
		const fileContent = `
			import modules from "./non-exist/*.js?glob&eager";
			console.log(modules);
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});

	it.skipIf(!supportImportAttributes)(
		"should eagerly load JSON modules with '?glob&eager' and import attributes",
		async () => {
			const fileContent = `
			import modules from "./json/*.json?glob&eager" with { type: "json" };
			console.log(modules);
		`;
			await fs.writeFile(mainPath, fileContent);

			const { stdout } = await exec(
				`node --import="${registerUrl}" "${mainPath}"`,
			);
			expect(stdout.toString()).toMatchSnapshot();
		},
	);

	it("should eagerly load parent ('../') modules with '?glob&eager'", async () => {
		mainPath = path.resolve(process.cwd(), "tests/fixtures/parent/sub/main.js");
		const fileContent = `
			import modules from "../*.js?glob&eager";
			console.log(modules);
		`;
		await fs.writeFile(mainPath, fileContent);

		const { stdout } = await exec(
			`node --import="${registerUrl}" "${mainPath}"`,
		);
		expect(stdout.toString()).toMatchSnapshot();
	});
});
