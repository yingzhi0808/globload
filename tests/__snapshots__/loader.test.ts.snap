// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Glob Loader > should dynamic import() eagerly with '?glob&eager' 1`] = `
"{
  'tests/fixtures/basic/product.js': [Module: null prototype] {
    default: 'electronics',
    price: 1200,
    productName: 'Laptop'
  },
  'tests/fixtures/basic/user.js': [Module: null prototype] {
    default: { role: 'user' },
    getAge: [Function: getAge],
    name: 'Alice'
  }
}
"
`;

exports[`Glob Loader > should dynamic import() lazily with '?glob' 1`] = `
"{
  'tests/fixtures/basic/product.js': [Function: tests/fixtures/basic/product.js],
  'tests/fixtures/basic/user.js': [Function: tests/fixtures/basic/user.js]
}
[Module: null prototype] {
  default: 'electronics',
  price: 1200,
  productName: 'Laptop'
}
[Module: null prototype] {
  default: { role: 'user' },
  getAge: [Function: getAge],
  name: 'Alice'
}
"
`;

exports[`Glob Loader > should eagerly load empty directory as empty object with '?glob&eager' 1`] = `
"{}
"
`;

exports[`Glob Loader > should eagerly load nested (**) modules with '?glob&eager' 1`] = `
"{
  'tests/fixtures/nested/data.js': [Module: null prototype] { a: 'a', b: 'b', c: 'c' },
  'tests/fixtures/nested/deep/product.js': [Module: null prototype] {
    default: 'electronics',
    price: 1200,
    productName: 'Laptop'
  },
  'tests/fixtures/nested/deep/user.js': [Module: null prototype] {
    default: { role: 'user' },
    getAge: [Function: getAge],
    name: 'Alice'
  }
}
"
`;

exports[`Glob Loader > should eagerly load non-existent path as empty object with '?glob&eager' 1`] = `
"{}
"
`;

exports[`Glob Loader > should eagerly load parent ('../') modules with '?glob&eager' 1`] = `
"{
  'tests/fixtures/parent/product.js': [Module: null prototype] {
    default: 'electronics',
    price: 1200,
    productName: 'Laptop'
  },
  'tests/fixtures/parent/user.js': [Module: null prototype] {
    default: { role: 'user' },
    getAge: [Function: getAge],
    name: 'Alice'
  }
}
"
`;

exports[`Glob Loader > should static import eagerly with '?glob&eager' 1`] = `
"{
  'tests/fixtures/basic/product.js': [Module: null prototype] {
    default: 'electronics',
    price: 1200,
    productName: 'Laptop'
  },
  'tests/fixtures/basic/user.js': [Module: null prototype] {
    default: { role: 'user' },
    getAge: [Function: getAge],
    name: 'Alice'
  }
}
"
`;

exports[`Glob Loader > should static import lazily with '?glob' 1`] = `
"{
  'tests/fixtures/basic/product.js': [Function: tests/fixtures/basic/product.js],
  'tests/fixtures/basic/user.js': [Function: tests/fixtures/basic/user.js]
}
[Module: null prototype] {
  default: 'electronics',
  price: 1200,
  productName: 'Laptop'
}
[Module: null prototype] {
  default: { role: 'user' },
  getAge: [Function: getAge],
  name: 'Alice'
}
"
`;
